gui/src/ruler.cc

```diff
@@ -0,0 +1,362 @@
+#include <cmath>
+#include <iostream>
+#include "ruler.hh"
+
+////////////////////////////////////////////////////////////////////////
+// Ruler
+
+Ruler::Ptr Ruler::create(Ruler::Orientation orientation, GtkWidget *drawingArea)
+{
+    Ruler::Ptr ruler{new Ruler(orientation)};
+    if (drawingArea != nullptr)
+    {
+        ruler->setDrawingArea(drawingArea);
+    }
+    ruler->update();
+    return ruler;
+}
+
+Ruler::Ruler(Ruler::Orientation orientation)
+        : orientation{orientation}
+{
+}
+
+Ruler::~Ruler()
+{
+    unregisterDrawingArea();
+}
+
+void Ruler::setDrawingArea(GtkWidget *widget)
+{
+    unregisterDrawingArea();
+    drawingArea = widget;
+
+    width = gtk_widget_get_allocated_width(widget);
+    height = gtk_widget_get_allocated_height(widget);
+
+    // Register callbacks
+    g_signal_connect(widget, "draw", G_CALLBACK(drawCallback), this);
+    g_signal_connect(widget, "size-allocate", G_CALLBACK(sizeAllocateCallback), this);
+}
+
+void Ruler::unregisterDrawingArea()
+{
+    if (drawingArea == nullptr) { return; }
+
+    g_signal_handlers_disconnect_by_data(drawingArea, this);
+    drawingArea = nullptr;
+}
+
+void Ruler::setRange(double lower, double upper)
+{
+    lowerLimit = lower;
+    upperLimit = upper;
+
+    if (drawingArea == nullptr) { return; }
+
+    update();
+
+    // We need to manually trigger the widget to redraw
+    gtk_widget_queue_draw(drawingArea);
+}
+
+void Ruler::update()
+{
+    if (drawingArea == nullptr) { return; }
+
+    // We need to calculate the distance between the largest ticks on the ruler
+    // We will try each interval sequentially until we find an interval which
+    // will produce segments of a large enough width/height when drawn
+
+    // Once we have the distance between the major ticks (the size of the segments)
+    // each segment is then divided into 5 parts assuming there's enough space
+    // and each of those parts is again divided into 2, again, assuming there's enough space
+
+    // Index in the ruler's VALID_INTERVALS array
+    int intervalIndex = 0;
+    // Each interval is multiplied by 10 raised to the power n
+    const int INTERVAL_BASE = 10;
+    int intervalN = 0;
+
+    while (true)
+    {
+        if (VALID_INTERVALS.at(intervalIndex) != 1 || intervalN != 0)
+        {
+            majorInterval = VALID_INTERVALS.at(intervalIndex) * pow(INTERVAL_BASE, intervalN);
+
+            // We check versus the width or height depending on orientation
+            double rulerSize = NAN;
```

Comment:
`rulerSize` can be computed before the loop, right? And if you write it as `rulerSize = (orientation==HORIZONTAL)?width:height`, then it can also be `const`